hydra-node/src/Hydra/HeadLogic.hs

```diff
@@ -646,35 +656,67 @@ onOpenNetworkReqTx env ledger st ttl tx =
   case applyTransactions currentSlot seenUTxO [tx] of
     Left (_, err)
       | ttl <= 0 ->
-        NewState (Open st{coordinatedHeadState = untrackTxInState})
-          `Combined` Effects [ClientEffect $ TxInvalid headId seenUTxO tx err]
+          NewState (Open st{coordinatedHeadState = untrackTxInState})
+            `Combined` Effects [ClientEffect $ TxInvalid headId seenUTxO tx err]
       | otherwise ->
           NewState (Open st{coordinatedHeadState = trackTxInState})
             `Combined` Wait (WaitOnNotApplicableTx err)
     Right utxo' ->
-      NewState
-        ( Open
-            st
-              { coordinatedHeadState =
-                  trackTxInState
-                    { seenTxs = seenTxs <> [tx]
-                    , seenUTxO = utxo'
+      Effects [ClientEffect $ TxValid headId tx]
+        `Combined` if isLeader parameters party nextSn && not snapshotInFlight
```

Comment:
On master, the `seenSnapshot` function is

```
-- | Get the last seen snapshot number given a 'SeenSnapshot'.
seenSnapshotNumber :: SeenSnapshot tx -> SnapshotNumber
seenSnapshotNumber = \case
  NoSeenSnapshot -> 0
  LastSeenSnapshot{lastSeen} -> lastSeen
  RequestedSnapshot{lastSeen} -> lastSeen
  SeenSnapshot{snapshot = Snapshot{number}} -> number
```

For this change proposal to work, we would need to change this function to be:

```
-- | Get the last seen snapshot number given a 'SeenSnapshot'.
seenSnapshotNumber :: SeenSnapshot tx -> SnapshotNumber
seenSnapshotNumber = \case
  NoSeenSnapshot -> 0
  LastSeenSnapshot{lastSeen} -> lastSeen
  RequestedSnapshot{requested} -> requested
  SeenSnapshot{snapshot = Snapshot{number}} -> number
```

But this breaks the code because we also use this function in `onOpenNetworkReqSn` where we really want the master's way of computing the seenSnapshot number. For instance, we can see that the following test becomes red:

```
  test/Hydra/NodeSpec.hs:69:7: 
  1) Hydra.Node emits a single ReqSn and AckSn as leader, even after multiple ReqTxs
       expected: [ReqSn {snapshotNumber = UnsafeSnapshotNumber 1, transactionIds = [1]},AckSn {signed = HydraSignature "4d5e417b001974f93c8932ec29efb0c93c75de211c5413d78b806e6dbc41606d10c1c64735bb384b8b3ff2165ed46dbba9d9593a80f102285a8d9892e3ffe609", snapshotNumber = UnsafeSnapshotNumber 1}]
        but got: [ReqSn {snapshotNumber = UnsafeSnapshotNumber 1, transactionIds = [1]}]
```

Since this change is not part of this P.R. original scope, I say we don't do it here and create a red bin out of this comment to figure out what's going on with this seen snapshot concept here.

Reply:
See #982